### PR TITLE
Creating MSigDB hetnet

### PR DESCRIPTION
**Pull request by gwaygenomics**
Modifying https://github.com/dhimmel/integrate to generate hetnet for MSigDB genesets

### CODE REVIEW

File: scripts/nbconverted/integrate-msigdb.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+    
+    # metanodes
+    'Gene': 'G',
+
+    'Cancer-Hallmarks': 'H',
+    'Positional-Gene-Sets': 'C1',
+    'Curated-Gene-Sets': 'C2',
+    'Motif-Gene-Sets': 'C3',
+    'Computational-Gene-Sets': 'C4',
+    'GO-Gene-Sets': 'C5',
```
Comment (dhimmel) [inline scripts/nbconverted/integrate-msigdb.py]: You may want to split C5 into three metanodes (one for each GO Domain). We've found cellular components are less informative of micro-data-bio-logy than MFs or BPs.

File: scripts/nbconverted/integrate-msigdb.py [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+
+
+# ## Gene Nodes
+
+# In[8]:
+
+
+# Load curated gene names from versioned resource 
+commit = 'a7362748a34211e5df6f2d185bb3246279760546'
+url = 'https://raw.githubusercontent.com/dhimmel/entrez-gene/{}/data/genes-human.tsv'.format(commit)
```
Comment (dhimmel) [inline scripts/nbconverted/integrate-msigdb.py]: I'd use [genes](https://github.com/cognoma/genes/blob/master/data/genes.tsv) from https://github.com/cognoma/genes and consider applying the [updater mapping](https://github.com/cognoma/genes/blob/master/data/updater.tsv). cognoma/genes is more up to date and the mapping is nice.

File: scripts/nbconverted/integrate-msigdb.py [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+    Output:
+    Adds to current graph; Returns the amount of filtered genesets
+    """
+    
+    # Build meta data dictionary to store node info
+    meta_data = {'license': license, 'source': collection_source}
+    
+    # Open the .gmt file and process each geneset
+    filtered_genesets = []
+    with open(collection_file, 'r') as collection_fh:
```
Comment (dhimmel) [inline scripts/nbconverted/integrate-msigdb.py]: Is nice to separate out GMT reading into a generative fxn like:

```python
def read_gmt(path):
    """Generate rows from a .gmt file."""
    read_file = open(path)
    reader = csv.reader(read_file, delimiter='\t')
    for row in reader:
        name = row[0]
        description = row[1]
        genes = set(row[2:])
        yield name, description, genes
    read_file.close()
```

But change that fxn to use `with` for file open. Not sure what I was thinking not using context management.

File: scripts/nbconverted/integrate-msigdb.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+            current_graph.add_node(kind=collection_kind, identifier=geneset_name, data=meta_data)
+
+            # Loop through all genes and add to the graph it should be considered
+            for gene in genes:
+                if gene in gene_list:
+                    source_id = ('Gene', gene)
+                    target_id = (collection_kind, geneset_name)
+                    
+                    edge_data = meta_data.copy()
+                    edge_data['unbiased'] = False
```
Comment (dhimmel) [inline scripts/nbconverted/integrate-msigdb.py]: No need for you to set this probably.

File: scripts/nbconverted/integrate-msigdb.py [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+# In[29]:
+
+
+get_ipython().run_cell_magic('time', '', "# Write nodes to a table\nnodes_file = os.path.join('hetnets', 'msigdb_nodes.tsv')\nhetio.readwrite.write_nodetable(graph, nodes_file)\n\n# Write edges to a table\nedges_file = os.path.join('hetnets', 'msigdb_edges.sif.gz')\nhetio.readwrite.write_sif(graph, edges_file)")
+
+
+# In[30]:
+
+
+get_ipython().run_cell_magic('time', '', "# Write a subset of edges to a table\nedges_file = os.path.join('hetnets', 'msigdb_edges-10.sif.gz')\nhetio.readwrite.write_sif(graph, edges_file, max_edges=10)\n\nedges_file = os.path.join('hetnets', 'msigdb_edges-5k.sif.gz')\nhetio.readwrite.write_sif(graph, edges_file, max_edges=5000)")
```
Comment (dhimmel) [inline scripts/nbconverted/integrate-msigdb.py]: You can probably delete these subsets... we just used them for cytoscape viz.

File: scripts/nbconverted/integrate-msigdb.py (hunk 4 above)
Comment (gwaygenomics) [reply scripts/nbconverted/integrate-msigdb.py]: This was the line I had a question about

Comment (gwaygenomics) [reply scripts/nbconverted/integrate-msigdb.py]: will remove

**Comment by gwaygenomics** [comment]
@dhimmel - do you think the output files in `hetnets/` are good to add in this PR?

**Comment by dhimmel** [comment]
Yeah. I'd add all data files that don't have a licensing issue. Probably using LFS?

**Comment by gwaygenomics** [comment]
>  I'd add all data files that don't have a licensing issue.

I believe none have this issue. See http://software.broadinstitute.org/gsea/msigdb_license_terms.jsp

File: scripts/nbconverted/integrate-msigdb.py (hunk 1 above)
Comment (gwaygenomics) [reply scripts/nbconverted/integrate-msigdb.py]: Sounds good. I think splitting out other sub-collections would work well too.

File: scripts/nbconverted/integrate-msigdb.py (hunk 2 above)
Comment (gwaygenomics) [reply scripts/nbconverted/integrate-msigdb.py]: I will use cognoma genes

File: scripts/nbconverted/integrate-msigdb.py (hunk 3 above)
Comment (gwaygenomics) [reply scripts/nbconverted/integrate-msigdb.py]: I think its easier to read the way I have it. It also takes only seconds to run

**Comment by gwaygenomics** [comment]
Thanks for comments @dhimmel - I made changes and/or responded to your comments above. The main updates include separating out several sub-collections. I think downstream applications will be easier to track with this level of detail. I also change the genes resource, git lfs track the hetnets, add results, and include an `environment.yml`. Ready for re-review!

File: environment.yml [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+- conda-forge::python=3.6
+- conda-forge::pandas=0.22.0
+- conda-forge::numpy=1.14.2
+- conda-forge::seaborn=0.8.1
+- conda-forge::jupyter=1.0.0
+- conda-forge::ipykernel=4.8.1
+- conda-forge::nb_conda_kernels=2.1.0
+- conda-forge::openpyxl=2.5.1
+- pip:
+  - hetio==0.2.3
```
Comment (dhimmel) [inline environment.yml]: I'd update hetio to 0.2.8. it should have several new features and bug fixes since 0.2.3

Comment (gwaygenomics) [reply environment.yml]: awesome, I will update accordingly

**Comment by gwaygenomics** [comment]
Thanks again @dhimmel - updated to `hetio=0.2.8`

**Pull request merged**